File: backend/agents/deployer.py
import os
from datetime import datetime, timezone
import uuid
import hashlib

class DeployerAgent:
    def __init__(self, db, manager):
        self.db = db
        self.manager = manager

    async def deploy(self, task_id: str, code: str, project_id: str):
        await self._log(task_id, "Deployer", "🚀 Starting deployment process...")
        
        try:
            # Simulate deployment
            commit_sha = hashlib.sha256(code.encode()).hexdigest()[:12]
            deployment_url = f"https://catalyst-{project_id[:8]}.deploy.catalyst.ai"
            
            await self._log(task_id, "Deployer", "📦 Building application...")
            await self._log(task_id, "Deployer", "☁️ Deploying to cloud...")
            
            # Create deployment record
            deployment_doc = {
                "id": str(uuid.uuid4()),
                "task_id": task_id,
                "url": deployment_url,
                "commit_sha": commit_sha,
                "cost": 0.25,
                "report": f"Deployment successful\nURL: {deployment_url}\nCommit: {commit_sha}\nStatus: Live",
                "created_at": datetime.now(timezone.utc).isoformat()
            }
            await self.db.deployments.insert_one(deployment_doc)
            
            await self._log(task_id, "Deployer", f"✅ Deployment successful: {deployment_url}")
            
            return {
                "deployment_url": deployment_url,
                "commit_sha": commit_sha,
                "status": "success"
            }
        except Exception as e:
            await self._log(task_id, "Deployer", f"❌ Deployment failed: {str(e)}")
            return {"deployment_url": "", "commit_sha": "", "status": "failed", "error": str(e)}

    async def _log(self, task_id: str, agent_name: str, message: str):
        log_doc = {
            "task_id": task_id,
            "agent_name": agent_name,
            "message": message,
            "timestamp": datetime.now(timezone.utc).isoformat()
        }
        await self.db.agent_logs.insert_one(log_doc)
        await self.manager.send_log(task_id, log_doc)

File: backend-go/internal/agents/deployer.go
package agents

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"time"

	"github.com/catalyst/backend/internal/config"
	"github.com/catalyst/backend/internal/database"
	"github.com/catalyst/backend/internal/models"
	"github.com/catalyst/backend/internal/websocket"
	"github.com/google/uuid"
)

type DeployerAgent struct {
	db        *database.Database
	wsManager *websocket.Manager
	cfg       *config.Config
}

func NewDeployerAgent(db *database.Database, wsManager *websocket.Manager, cfg *config.Config) *DeployerAgent {
	return &DeployerAgent{
		db:        db,
		wsManager: wsManager,
		cfg:       cfg,
	}
}

func (a *DeployerAgent) Deploy(taskID, code, projectID string) (string, string, error) {
	a.log(taskID, "Deployer", "🚀 Starting deployment process...")

	// Generate commit SHA
	hash := sha256.Sum256([]byte(code))
	commitSHA := hex.EncodeToString(hash[:])[:12]

	// Generate deployment URL
	deploymentURL := fmt.Sprintf("https://catalyst-%s.deploy.catalyst.ai", projectID[:8])

	a.log(taskID, "Deployer", "📦 Building application...")
	time.Sleep(1 * time.Second)

	a.log(taskID, "Deployer", "☁️ Deploying to cloud...")
	time.Sleep(1 * time.Second)

	// Create deployment record
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	deployment := models.Deployment{
		ID:        uuid.New().String(),
		TaskID:    taskID,
		URL:       deploymentURL,
		CommitSHA: commitSHA,
		Cost:      0.25,
		Report:    fmt.Sprintf("Deployment successful\nURL: %s\nCommit: %s\nStatus: Live", deploymentURL, commitSHA),
		CreatedAt: time.Now(),
	}

	a.db.Deploys.InsertOne(ctx, deployment)

	a.log(taskID, "Deployer", fmt.Sprintf("✅ Deployment successful: %s", deploymentURL))

	return deploymentURL, commitSHA, nil
}

func (a *DeployerAgent) log(taskID, agentName, message string) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	log := models.AgentLog{
		ID:        uuid.New().String(),
		TaskID:    taskID,
		AgentName: agentName,
		Message:   message,
		Timestamp: time.Now(),
	}

	a.db.Logs.InsertOne(ctx, log)
	a.wsManager.SendLog(taskID, map[string]interface{}{
		"task_id":    taskID,
		"agent_name": agentName,
		"message":    message,
		"timestamp":  log.Timestamp.Format(time.RFC3339),
	})
}